toolchain/diagnostics/diagnostic.h

```diff
@@ -58,10 +58,11 @@ struct DiagnosticLoc {
   llvm::StringRef filename;
   // A reference to the line of the error.
   llvm::StringRef line;
-  // 1-based line number.
-  int32_t line_number = -1;
+  // 1-based line number. This defaults to 0 because LLVM's DebugInfo uses that
+  // for unknown values, and we're sharing the logic.
+  int32_t line_number = 0;
```

Comment:
I certainly don't mind mapping -1 -> 0 in the debug info lowering, rather than sharing this representational detail.

Though independently of that, I'd favor 0 as the invalid value (& maybe use an unsigned integer here, though that's probably more contentious) if it's not an otherwise valid value - then it's clear there's only one invalid value (0) and all other values are valid. Rather than wondering if 0 encodes something different from -1 (or -2), etc.

Reply:
FWIW, I'm really tempted to split logic just to retain -1.

The result of `-1` as `line: 4294967295` gave a strong hint as to which code was both receiving and failing to handle the invalid value. I don't think `0` will do as well, just because it's a more reasonable unsigned value, and blends in better. Also, `-1` is more familiar to me as a common, invalid value.

Reply:
I'm good with that - though I'd like it to be documented that the invalid value is -1, positive values are expected, and nothing else is expected. And having checks == -1 exactly, rather than <= 0 - the latter makes me a bit uncomfortable "what values are there?". -1 is clearer for me to read.

Reply:
Per discussion, switched to -1 and renamed GetDiagnosticLoc to GetLocForDI